### PR TITLE
Introduce EtcdDtabStore

### PR DESCRIPTION
**Pull request by adleong**
Add a DtabStore which stores dtabs in etcd using the Etcd client.

### CODE REVIEW

File: namerd/storage/etcd/src/main/scala/io/buoyant/namerd/storage/etcd.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import com.twitter.finagle.{Http, Path}
+import io.buoyant.config.types.Port
+import io.buoyant.etcd.Key
+import io.buoyant.namerd.storage.etcd.EtcdDtabStore
+import io.buoyant.namerd.{DtabStore, DtabStoreConfig, DtabStoreInitializer}
+
+case class EtcdConfig(
+  host: Option[String],
+  port: Option[Port],
+  key: Path
```
Comment (olix0r) [inline namerd/storage/etcd/src/main/scala/io/buoyant/namerd/storage/etcd.scala]: should this be `root` or something?  should it have a default?

File: namerd/storage/etcd/src/main/scala/io/buoyant/namerd/storage/etcd/EtcdDtabStore.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                updates.update(Activity.Failed(new IllegalStateException(s"${key.path.show} is not a data node")))
+
+            }
+          case _ =>
+        }
+
+        case Throw(ApiError(ApiError.KeyNotFound, _, _, _)) =>
+          updates.update(Activity.Ok(None))
+        case Throw(e) =>
+          updates.update(Activity.Failed(e))
```
Comment (olix0r) [inline namerd/storage/etcd/src/main/scala/io/buoyant/namerd/storage/etcd/EtcdDtabStore.scala]: fwiw, the `states() = Activity...` form is fairly idiomatic throughout the rest of our codebase, as well as finagle's...  i don't feel strongly, though

File: project/LinkerdBuild.scala [hunk 3]
```diff
@@ -265,7 +270,7 @@ object LinkerdBuild extends Base {
       // Bundle includes all of the supported features:
       .configDependsOn(Bundle)(
         Namer.consul, Namer.k8s, Namer.marathon, Namer.serversets,
-        Storage.k8s, Storage.zk
+        Storage.k8s, Storage.zk, Storage.etcd, Storage.inMemory
```
Comment (olix0r) [inline project/LinkerdBuild.scala]: alphasort?

File: namerd/storage/etcd/src/main/scala/io/buoyant/namerd/storage/etcd.scala [hunk 4]
```diff
@@ -10,14 +10,13 @@ import io.buoyant.namerd.{DtabStore, DtabStoreConfig, DtabStoreInitializer}
...
-  key: Path
+  pathPrefix: Option[Path]
 ) extends DtabStoreConfig {
   import EtcdConfig._
-  require(key != null, "'key' field is required")
   @JsonIgnore
   override def mkDtabStore: DtabStore = {
     new EtcdDtabStore(new Key(
-      key,
+      pathPrefix.getOrElse(Path.read("/dtabs")),
```
Comment (olix0r) [inline namerd/storage/etcd/src/main/scala/io/buoyant/namerd/storage/etcd.scala]: nit: let's use `/namerd/dtabs` so we have a place to stick other non-dtab things later.

**Comment by olix0r** [comment]
lgtm, thanks!

File: namerd/storage/etcd/src/integration/scala/io/buoyant/namerd/storage/etcd/EtcdDtabStoreIntegrationTest.scala [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  test("observe follows changes") { etcd =>
+    val store = mkStore(etcd)
+    val obs = store.observe("hello")
+    assert(await(obs.toFuture).isEmpty)
+    await(store.create("hello", Dtab.read("/hello => /world")))
+    assert(await(obs.toFuture).get.dtab == Dtab.read("/hello => /world"))
+    val version = await(obs.toFuture).get.version
+    await(store.update("hello", Dtab.read("/goodbye => /world"), version))
+    assert(await(obs.toFuture).get.dtab == Dtab.read("/goodbye => /world"))
+  }
```
Comment (klingerf) [inline namerd/storage/etcd/src/integration/scala/io/buoyant/namerd/storage/etcd/EtcdDtabStoreIntegrationTest.scala]: TIOLI: would it be worth validating that the version changes after each update as part of this test?

**Comment by klingerf** [comment]
⭐ 🐑

**Pull request merged**